plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RewriteCharVarcharComparison.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.redshift;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.plugin.base.expression.ConnectorExpressionRule;
+import io.trino.plugin.jdbc.QueryParameter;
+import io.trino.plugin.jdbc.expression.ComparisonOperator;
+import io.trino.plugin.jdbc.expression.ParameterizedExpression;
+import io.trino.spi.expression.Call;
+import io.trino.spi.expression.Constant;
+import io.trino.spi.expression.FunctionName;
+import io.trino.spi.expression.Variable;
+import io.trino.spi.type.CharType;
+import io.trino.spi.type.VarcharType;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.constant;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.type;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.variable;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static java.lang.String.format;
+
+public class RewriteCharVarcharComparison
+        implements ConnectorExpressionRule<Call, ParameterizedExpression>
+{
+    private static final Capture<Variable> LEFT_VARIABLE = newCapture();
+
+    private static final Capture<Variable> RIGHT_VARIABLE = newCapture();
+    private static final Capture<Constant> RIGHT_CONSTANT = newCapture();
+
+    private final Pattern<Call> pattern;
+
+    public RewriteCharVarcharComparison(Set<ComparisonOperator> enabledOperators)
+    {
+        Set<FunctionName> functionNames = enabledOperators.stream()
+                .map(ComparisonOperator::getFunctionName)
+                .collect(toImmutableSet());
+
+        pattern = call()
+                .with(functionName().matching(functionNames::contains))
+                .with(type().equalTo(BOOLEAN))
+                .with(argumentCount().equalTo(2))
+                .with(argument(0).matching(variable().capturedAs(LEFT_VARIABLE).with(type().matching(type -> type instanceof VarcharType || type instanceof CharType))))
+                .or(
```

Comment:
Could we collapse this to have single `or()` like this?

```java
                .or(
                        inner -> inner.with(argument(1).matching(variable().capturedAs(RIGHT_VARIABLE)
                                .with(type().matching(type -> type instanceof VarcharType || type instanceof CharType)))),
                        inner -> inner.with(argument(1).matching(constant().capturedAs(RIGHT_CONSTANT)
                                .with(type().matching(type -> type instanceof VarcharType || type instanceof CharType)))));
```

Reply:
We'll discuss further in https://github.com/trinodb/trino/issues/23922